contracts/test/AllowanceTarget.t.sol

```diff
@@ -0,0 +1,134 @@
+        address mySpender = address(new MockStrategy());
+        vm.expectRevert("AllowanceTarget: SetSpender in progress");
+        allowanceTarget.setSpenderWithTimelock(mySpender);
+    }
+
+    function testSetSpenderWithTimelock() public {
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+        assertEq(allowanceTarget.newSpender(), newSpender);
+        assertEq(allowanceTarget.timelockExpirationTime(), block.timestamp + 1 days);
+    }
+
+    /*********************************
+     *  test: setSpenderWithTimelock *
+     *********************************/
+
+    function testCantCompleteSetSpenderBeforeSet() public {
+        vm.expectRevert("AllowanceTarget: no pending SetSpender");
+        allowanceTarget.completeSetSpender();
+    }
+
+    function testCantCompleteSetSpenderTooEarly() public {
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+        vm.expectRevert("AllowanceTarget: time lock not expired yet");
+        allowanceTarget.completeSetSpender();
+    }
+
+    function testCompleteSetSpender() public {
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+        // fast forward
+        vm.warp(allowanceTarget.timelockExpirationTime());
+        allowanceTarget.completeSetSpender();
+    }
+
+    /*********************************
+     *         test: teardown        *
+     *********************************/
+
+    function testCantTeardownIfNotSpender() public {
+        vm.prank(bob);
+        vm.expectRevert("AllowanceTarget: not the spender");
+        allowanceTarget.teardown();
+    }
+
+    function testTeardown() public {
+        // Seems like the test contract itself have issue with receving ether.
+        // So set spender to mock contract first.
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+        // fast forward
+        vm.warp(allowanceTarget.timelockExpirationTime());
+        allowanceTarget.completeSetSpender();
+
+        BalanceSnapshot.Snapshot memory beneficiary = BalanceSnapshot.take(address(newSpender), Addresses.ETH_ADDRESS);
+        uint256 heritage = 10 ether;
+        vm.deal(address(allowanceTarget), heritage);
+        vm.prank(newSpender);
+        allowanceTarget.teardown();
+        beneficiary.assertChange(int256(heritage));
+    }
+
+    /*********************************
+     *         test: executeCall     *
+     *********************************/
+
+    function testCantExecuteCallIfNotSpender() public {
+        vm.prank(bob);
+        vm.expectRevert("AllowanceTarget: not the spender");
+
+        bytes memory data;
+        allowanceTarget.executeCall(payable(address(newSpender)), data);
+    }
+
+    function testExecuteCall() public {
+        MockERC20 token = new MockERC20("Test", "TST", 18);
+
+        // mint(address to, uint256 value)
+        allowanceTarget.executeCall(payable(address(token)), abi.encodeWithSelector(MockERC20.mint.selector, address(bob), 1e18));
```

Comment:
Should we check bob balance to ensure allowanceTarget executing the call we expected?

Reply:
Added in 219124dce1aacf31de59a238fe02e114aaab9e9e